onnx/version_converter/helper.cc

```diff
@@ -0,0 +1,79 @@
+// Helper Methods for Adapters
+
+#include "onnx/version_converter/helper.h"
+
+namespace ONNX_NAMESPACE { namespace version_conversion {
+    bool onnx_opset7_requires_broadcasting(int64_t axis,
+        const std::vector<Dimension>& input1_sizes,
+        const std::vector<Dimension>& input2_sizes) {
+      // Assert that broadcasting semantics are correct
+      for (int i = 0; i < (int) input2_sizes.size(); i++) {
+        ONNX_ASSERTM(input2_sizes[i].dim == input1_sizes[(int) axis + i].dim ||
+            input2_sizes[i].dim == 1, "Dimension %d of input 2 does not match "
+            "dimension %d of input 1 or the value 1", i, (int) axis + i);
+      }
+      // Return false if a Reshape is necessary for forward compatibility
+      return axis != (int) (input1_sizes.size() - input2_sizes.size());
+    }
+
+    bool numpy_unibroadcastable(const std::vector<Dimension>& input1_sizes,
+        const std::vector<Dimension>& input2_sizes) {
+      // Assert that equal of input1 larger
+      ONNX_ASSERTM(input1_sizes.size() >= input2_sizes.size(),
+          "ONNX Opset 6 can only broadcast input 2 to input 1");
+      // Assert that axis is input1_sizes.size()-input2_sizes.size()
+      bool broadcast = false;
+      int axis = (int) (input1_sizes.size() - input2_sizes.size());
+      for (int i = 0; i < (int) input2_sizes.size(); i++) {
+        ONNX_ASSERTM(input2_sizes[i].dim == input1_sizes[axis + i].dim ||
+            input2_sizes[i].dim == 1, "Dimension %d of input 2 does not match "
```

Comment:
You can return int as result, like 0 is no broadcast needed, 1 is broadcast is needed. -1 is non-broadcastable.

And give better error message in the caller.

Reply:
The reason we don't distinguish between non-broadcastable and no broadcast needed/broadcast needed is that the asserts will interrupt whichever program if the inputs are non-broadcastable, so we'll never reach any such return value.

Where would the enum be useful here?